hwaccel and fifo improvements

galad87 opened the pull request:
Here's some changes I had around since last year.
It adds a way to wrap AVFrame inside hb_buffer, at the moment is enabled only for the hardware frames used by NVDEC.
I added a new "storage" field to hb_buffer, to make it easy to use it as a wrapper around different storage types, for example CMSampleBuffer to avoid a memcpy in encvt.c.

**Test on:**

- [ ] Windows 10+  (via MinGW)
- [ ] macOS 10.13+
- [ ] Ubuntu Linux

sr55 commented:
- Suspect Picture previews may be broken.  (Segfault)
- NVEnc on it's own is fine.
- NVDec is indeed broken.  Looks like a segfault.

Looks like 5~7% performance improvement on a few different types of encodes.

galad87 commented:
Did you test it using BBB with align a/v on?
I refactored the common hwaccel code to test it on macOS, and it actually works (with some small issues). Anyway can you try again if NVDec is better now or worse?

sr55 commented:
NVDec is fixed.  Both with and without AlignAV on BBB.

Preview is still a crash for now. When I get a bit more time I can get a stack trace if you need it.

theHamsta commented on libhb/encavcodec.c:
```diff
@@ -846,19 +842,13 @@ int encavcodecInit( hb_work_object_t * w, hb_job_t * job )
+        hb_hwaccel_hwframes_ctx_init(context, job);
     }
-#else
     context->pix_fmt = job->output_pix_fmt;
```
This works for me when I change it to

```c
    if (hb_hwaccel_is_full_hardware_pipeline_enabled(pv->job))
    {
        context->hw_device_ctx = pv->job->hw_device_ctx;
        hb_hwaccel_hwframes_ctx_init(context, job);
        context->pix_fmt = job->hw_pix_fmt;
    } else {
        context->pix_fmt = job->output_pix_fmt;
    }
```
otherwise the pix_fmt of the hardware ctx and the hardware frame ctx wouldn't match

galad87 replied:
Thanks, I'll change it.

theHamsta commented on libhb/handbrake/nvenc_common.h:
```diff
@@ -18,20 +18,4 @@ int            hb_nvenc_av1_available();
 int            hb_check_nvdec_available();
 
-
-char*          hb_map_nvenc_preset_name (const char * preset);
```
removing this will cause compilation to fail, since the function is still used in decavcodec.c/encavcodec.

galad87 replied:
Whoops.

theHamsta reviewed (commented):
This change looks great!

Nvidia can provide you Ada HW instances for the development of this feature or testing Nvidia support in general. Please let me know if you're interested or feel free to ping me when you want me to test something.

I'm still seeing refcounting issues when the hw_context is destroyed.

galad87 commented:
A way to test this on Nvidia would be useful indeed 😆 I'll make a note to contact you in the next days.
I should try to limit the scope of this PR and get this in, even if not perfect it will be better than the current situation.

theHamsta commented:
Once this is completed, adding support for 

E.g. enabling `scale_cuda` in `cropscale.c` would be just.
```c
        if(job && job->hw_pix_fmt == AV_PIX_FMT_CUDA) {
            hb_dict_set_int(avsettings, "w", width);
            hb_dict_set_int(avsettings, "h", height);
            hb_dict_set_string(avsettings, "interp_algo", "lanczos");
            hb_dict_set_string(avsettings, "format", av_get_pix_fmt_name(init->job->input_pix_fmt));
            hb_dict_set(avfilter, "scale_cuda", avsettings);
        } else if ((width % 2) == 0 && (height % 2) == 0 &&
```
"crop" works out of the box also with HW pix formats.

The only non-obvious change was that I had to set the hw frames context also to the buffer src in `hb_avfilter_graph_init`
```c
#if HB_PROJECT_FEATURE_NVENC
        if (hb_nvdec_is_enabled(init->job))
        {
            init->pix_fmt = AV_PIX_FMT_CUDA;
            par = av_buffersrc_parameters_alloc();
            par->format = init->pix_fmt;
            par->frame_rate.num = init->geometry.par.num;
            par->frame_rate.den = init->time_base.den;
            par->width = init->geometry.width;
            par->height = init->geometry.height;
            par->hw_frames_ctx = init_hw_frames_ctx((AVBufferRef*)init->job->hw_device_ctx,
                                                    init->job->hw_ctx,
                                                    par->width,
                                                    par->height);
            if(!par->hw_frames_ctx){
                goto fail;
            }
            par->sample_aspect_ratio.num = init->geometry.par.num;
            par->sample_aspect_ratio.den = init->geometry.par.den;
            par->time_base.num = init->time_base.num;
            par->time_base.den = init->time_base.den;
        }
#endif
        filter_args = hb_strdup_printf(
                    "width=%d:height=%d:pix_fmt=%d:sar=%d/%d:"
                    "time_base=%d/%d:frame_rate=%d/%d",
                    init->geometry.width, init->geometry.height, init->pix_fmt,
                    init->geometry.par.num, init->geometry.par.den,
                    init->time_base.num, init->time_base.den,
                    init->vrate.num, init->vrate.den);
    }
    avfilter = append_filter(graph, "buffer", filter_args, par);
    av_free(par);
    par = NULL;
    free(filter_args);
    filter_args = NULL;
```

sr55 commented:
**Minor thing:**
Before:  [13:54:03]    + decoder: h264_nvdec 8-bit (yuv420p)
After:    [13:57:50]    + decoder: h264_hwaccel 8-bit (yuv420p, (null))

We might want to note in the log that hardware scaling is being used.

sr55 commented:
Looks like Nvidia's scaler is running ~70-80fps slower than the software scaler. 
 
[AFTER-720p.txt](https://github.com/HandBrake/HandBrake/files/11373310/AFTER-720p.txt)   554.477234 fps
[BEFORE - 720p.txt](https://github.com/HandBrake/HandBrake/files/11373311/BEFORE.-.720p.txt)   623.315735 fps

No scaling is pretty much identical with NVDEC/NVENC in line .

galad87 commented:
I implemented the GPU to CPU decoder path too. But I forgot to update after the changes to the VFR filter, so it's always using the slow path, oops.

galad87 commented:
the "null" string  is printed when it's using the GPU to CPU path.
I committed a potential fix.

sr55 commented:
Just got a repeatable crash at encode complete. During muxing.  (With or without scaling) 

[crash - no scale.txt](https://github.com/HandBrake/HandBrake/files/11373650/crash.-.no.scale.txt)

Performance with scaling on GPU is improved significantly. (782.343628 fps vs 554.477234 fps  before  (or 623.315735 fps with software))

sr55 commented:
Another issue,  The scaling isn't doing what you think it is :)

![image](https://user-images.githubusercontent.com/628593/235681480-caf01a79-999e-459c-9b8b-93640f87886d.png)

galad87 commented:
😅 oh well.
I guess it will be a bit hard to fix until I get a way to test it.

sr55 commented:
Let me see if I can get a stack trace for it.

sr55 commented:
No joy getting a stack trace.  

`Encoding: task 1 of 1, 90.76 %[16:39:22] mux: track 0, 600 frames, 9523055 bytes, 7605.77 kbps, fifo 1024
warning: Critical error detected c0000374
--Type <RET> for more, q to quit, c to continue without paging--
Thread 23 received signal SIGTRAP, Trace/breakpoint trap.
[Switching to Thread 10256.0x6f40]
0x00007ffc55edbda3 in ?? ()
(gdb) bt
#0  0x00007ffc55edbda3 in ?? ()`






Did notice one other issue.

-i bbb.mp4 -o test.mp4 -e nvenc_h264 -a none --enable-hw-decoding=nvdec --stop-at seconds:10
Works

-i bbb.mp4 -o test.mp4 -e nvenc_h264 -a none  --stop-at seconds:10
(with or without --disable-hw-decoding)
fails


[16:38:24]    + decoder: h264_hwaccel 8-bit (yuv420p, (null))
,,,
[16:38:24] sync: expecting 660 video frames
[16:38:24] encavcodecInit: H.264 (Nvidia NVENC)
[16:38:24] encavcodec: encoding at rc=vbr, 22.00
ERROR: hwaccel: failed to initialize hw frames context
[16:38:24] encavcodec: encoding with stored aspect 1/1
[h264_nvenc @ 00000000584a2ec0] Specified pixel format -1 is invalid or not supported
[16:38:24] encavcodecInit: avcodec_open failed
ERROR: Failure to initialise thread 'FFMPEG encoder (libavcodec)'
[16:38:24] vfr: 0 frames output, 0 dropped
[16:38:24] vfr: lost time: 0 (0 frames)
[16:38:24] vfr: gained time: 0 (0 frames) (0 not accounted for)
[16:38:24] h264-decoder done: 0 frames, 0 decoder errors

galad87 commented:
@sr55 I think I fixed the crashes.

galad87 commented:
The scale filter works now, but it can only scale, it doesn't seem to support cropping, or I misconfigured something.

sr55 commented:
Not seeing any crashing anymore but it looks like it's switching to software decode path when downscaling.

theHamsta commented on libhb/encavcodec.c:
```diff
@@ -11,15 +11,11 @@
 #include "handbrake/nal_units.h"
 
-#if HB_PROJECT_FEATURE_NVENC
-#include "handbrake/nvenc_common.h"
```
for me, clang-17 will complain without this include, that it won't have a definition for `hb_map_nvenc_preset_name`
```
  : ../libhb/encavcodec.c:1537:22: error: call to undeclared function 'hb_map_nvenc_preset_name'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
  :             preset = hb_map_nvenc_preset_name(preset);
  :                      ^
  : ../libhb/encavcodec.c:1537:20: error: incompatible integer to pointer conversion assigning to 'const char *' from 'int' [-Wint-conversion]
  :             preset = hb_map_nvenc_preset_name(preset);
  :                    ^ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

galad87 committed:
libhb: generalize the hwaccel implementation.
Introduce a way to wrap AVFrame and CMSampleBuffer inside a hb_buffer_t to avoid unneeded conversions. The wrapping is enabled only for hw AVFrame at the moment.
Enable VideoToolbox hw decoders.

theHamsta commented:
At the moment, there is still problem when using the crop filter when scale_cuda comes after it. This seems to be caused by a bug in FFmpeg. The HW implementation of crop (https://github.com/FFmpeg/FFmpeg/blob/4006c71d19e77fbaacc5242fabf077cc84c9088c/libavfilter/vf_crop.c#L297-L302) just set additional `crop_XXX` fields to the frame, that are currently ignored by `scale_cuda` https://github.com/FFmpeg/FFmpeg/blob/4006c71d19e77fbaacc5242fabf077cc84c9088c/libavfilter/vf_scale_cuda.c#L503-L530 (it would need to offset it's writes and update the relevant height/width of the input frame). scale_vulkan is regarding those fields (https://github.com/FFmpeg/FFmpeg/blob/4006c71d19e77fbaacc5242fabf077cc84c9088c/libavfilter/vf_scale_vulkan.c#L118-L121). Is there are scale implementation for videotoolbox (I haven't found one yet)?

The software implementation of crop (https://github.com/FFmpeg/FFmpeg/blob/4006c71d19e77fbaacc5242fabf077cc84c9088c/libavfilter/vf_crop.c#L303-L322) that is currently doing pointer arithmetic might happen to work for the combination of crop/scale for CUDA, but due to misalignment and different representation of other HW implementations not in the general case.

galad87 commented:
I think I'll leave the crop issue to someone else, I disabled it again for now, but I fixed the filters in decavcodec.c.
Anyway, now it should be enough, some place are still clearly ugly, but it should work.

If someone could test this too and try to various 8 to 10bit and 10bit to 8 combinations.

galad87 commented:
@theHamsta is it be better to use a different av_hwframe_ctx for each filter or to share the same one between all the filters?

theHamsta commented:
> I think I'll leave the crop issue to someone else, I disabled it again for now, but I fixed the filters in decavcodec.c.
Anyway, now it should be enough, some place are still clearly ugly, but it should work.

You're right, this can be handled in a different PR. I just wanted to dump the info in some place where I could find it again.

> @theHamsta is it be better to use a different av_hwframe_ctx for each filter or to share the same one between all the filters?

I don't know FFmpeg well enough to answer that. But at first glance it looks that `HWFramesContext` just hold data describing the types of frames currently used https://github.com/FFmpeg/FFmpeg/blob/4006c71d19e77fbaacc5242fabf077cc84c9088c/libavutil/hwcontext.c#L248-L304 , also the private data of CUDA hw frames context is not too exciting https://github.com/FFmpeg/FFmpeg/blob/4006c71d19e77fbaacc5242fabf077cc84c9088c/libavutil/hwcontext_cuda.c#L33-L36. I'd assume that you would just set the relevant HWFramesCtx at the begining of the filter graph and it would make sense to me to have multiple of them when the format and height/width of the frames changes. The frames context seems to own the buffer pool to allocate the frames which also needs to be different when resolution changes.

:pray: Let me ask the true senseis: @rarzumanyan . This will finally use your `colorspace_cuda` filter.

Nomis101 commented:
> If someone could test this too and try to various 8 to 10bit and 10bit to 8 combinations.

Today I have tested some 8-to-10-bit and 10-to-8-bit conversions on my Intel Mac (macOS 13.4). AV1, H.264, H265, H265 (VideoToolbox). And a color space conversion. Couldn't see any issue.
